### PR TITLE
Fix/chat refactor

### PR DESCRIPTION
**Pull request by alexbojko**
Fix for issue CR-6

### Problem
is described in issue CR-6 in lira

### Solution
Refactoring

### CODE REVIEW

**Comment by cmltaWt0** [comment]
NOTE: `find . -name \*.pyc -delete` - pyc's should be deleted

File: mysite/ct/models.py [hunk 1]
```diff
@@ -372,6 +374,33 @@ def add_concept_link(self, concept, relationship, addedBy):
             return self.conceptlink_set.create(concept=concept, addedBy=addedBy,
                                                relationship=relationship)
 
+    def get_html(self, message, chat, **kwargs):
+        from chat.models import STATUS_OPTIONS
```
Comment (cmltaWt0) [inline mysite/ct/models.py]: Move it from method body please.

File: mysite/ct/models.py [hunk 2]
```diff
@@ -7,6 +7,8 @@
 from django.utils import timezone
 from django.core.urlresolvers import reverse
 from django.db.models import Q, Count, Max
+from django.utils.safestring import mark_safe
+from ct.templatetags.ct_extras import md2html
```
Comment (cmltaWt0) [inline mysite/ct/models.py:12]: Add blank line.

File: mysite/ct/models.py [hunk 3]
```diff
@@ -1016,6 +1053,49 @@ def get_next_step(self):
             if self.studenterror_set.count() == 0:
                 return self.CLASSIFY_STEP, 'classify your error(s)'
 
+    def get_html(self, message, chat, **kwargs):
+        from chat.models import EVAL_OPTIONS
```
Comment (cmltaWt0) [inline mysite/ct/models.py]: Same as above.

Comment (cmltaWt0) [inline mysite/ct/models.py]: Also why do we need t keep this in `chat` app.
Are this options belong to `ct` app?

File: mysite/ct/models.py [hunk 4]
```diff
@@ -1016,6 +1053,49 @@ def get_next_step(self):
...
+        html = message.text
+        if message.input_type == 'text':
+            html = mark_safe(md2html(self.text))
+        else:
+            CONF_CHOICES = dict(Response.CONF_CHOICES)
```
Comment (cmltaWt0) [inline mysite/ct/models.py:1074]: Same as above - where to place this.

File: mysite/fsm/base_fsm_node.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+from django.core.urlresolvers import reverse
+from django.utils.safestring import mark_safe
+from chat.models import Message, EVAL_OPTIONS, STATUS_OPTIONS
+from ct.models import UnitStatus, UnitLesson, Lesson, Response
+# from chat.models import Chat, Message, ChatDivider, UnitError
```
Comment (cmltaWt0) [inline mysite/fsm/base_fsm_node.py]: Remove commented out code.

File: mysite/fsm/base_fsm_node.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+from django.core.urlresolvers import reverse
+from django.utils.safestring import mark_safe
+from chat.models import Message, EVAL_OPTIONS, STATUS_OPTIONS
```
Comment (cmltaWt0) [inline mysite/fsm/base_fsm_node.py]: Add blank line.

File: mysite/fsm/base_fsm_node.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+from ct.models import UnitStatus, UnitLesson, Lesson, Response
+# from chat.models import Chat, Message, ChatDivider, UnitError
+import logging
+import re
```
Comment (cmltaWt0) [inline mysite/fsm/base_fsm_node.py:1]: Move to the very beginning of this module according to pep8.

File: mysite/fsm/base_fsm_node.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+from functools import partial
+from ct.templatetags.ct_extras import md2html
+
+
+class BaseFMSNode(object):
+    """
+    Base class for all FSM nodes which implements logic to get html (get_html method), get options (get_options)
```
Comment (cmltaWt0) [inline mysite/fsm/base_fsm_node.py]: Docstring formatting - do no exceed 80 chars please.

Comment (cmltaWt0) [reply mysite/fsm/base_fsm_node.py]: First line for the title. Then blank line. Then additional info.

File: mysite/fsm/base_fsm_node.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        new_msg = self._get_message(chat, message, current)
+        if not hasattr(self, '_get_message'):
+            if self.is_wait_node(self.name):
+                return self._get_wait_msg(chat, message, current)
+            logging.error("Node {} doesn't have _get_message method. Skipping it...".format(self.__class__.__name__))
+        if not new_msg:
+            return message
+        else:
+            return new_msg
```
Comment (cmltaWt0) [inline mysite/fsm/base_fsm_node.py:71]: Add blank line.

File: mysite/fsm/mixins.py [hunk 10]
```diff
@@ -153,287 +146,291 @@ def get_data_attr(self, attr):
...
-            lookup = dict(
-                chat=chat,
-                text=self.title,
-                kind='button',
-                is_additional=False,
-                owner=chat.user
-            )
-            message = Message.objects.get_or_create(**lookup)[0]
-        return message
+# class ChatMixin(object):
```
Comment (cmltaWt0) [inline mysite/fsm/mixins.py]: Please remove this.
We have diff in git.

File: mysite/fsm/models.py [hunk 11]
```diff
@@ -9,7 +9,7 @@
 
 from ct.ct_util import reverse_path_args
 from ct.models import Course, Role
-from mixins import JSONBlobMixin, ChatMixin
+from mixins import JSONBlobMixin #, ChatMixin
```
Comment (cmltaWt0) [inline mysite/fsm/models.py]: Remove this.

File: mysite/fsm/models.py [hunk 12]
```diff
@@ -123,7 +123,7 @@ def __set__(self, obj, val):
         raise AttributeError('read only attribute!')
 
 
-class FSMNode(JSONBlobMixin, ChatMixin, models.Model):
+class FSMNode(JSONBlobMixin, models.Model):  #ChatMixin
```
Comment (cmltaWt0) [inline mysite/fsm/models.py]: Remove this.

**Review by cmltaWt0 — CHANGES_REQUESTED** [review]

File: mysite/ct/models.py (hunk 1 above)
Comment (alexbojko) [reply mysite/ct/models.py]: it will completely broke the whole project because of circular imports there

File: mysite/ct/models.py (hunk 3 above)
Comment (alexbojko) [reply mysite/ct/models.py]: same as above

Comment (alexbojko) [reply mysite/ct/models.py]: good idea

File: mysite/ct/models.py (hunk 4 above)
Comment (alexbojko) [reply mysite/ct/models.py:1074]: not good idea. we are using CONF_CHOICES in place where it is defined. I don't think we should move it somewhere

File: mysite/fsm/base_fsm_node.py (hunk 8 above)
Comment (alexbojko) [reply mysite/fsm/base_fsm_node.py]: why not 120 chars? we all have wide displays and 80 symbols is very short string... I thought we are using 120 chars/line

**Pull request closed without merging**